File: api/routers/profiler.py

from fastapi import APIRouter, HTTPException, Query, Body
from api.services.profiler_service import ProfilerService
from api.services.data_loader import DATA_DIR
import json

router = APIRouter()

@router.get("/stats/profiler/{ticker}", tags=["Stats"])
async def get_profiler_stats(
    ticker: str, 
    days: int = Query(50, description="Number of days to analyze")
):
    """
    Get Profiler Statistics (Status, Broken) for the last N days.
    """
    result = ProfilerService.analyze_profiler_stats(ticker, days)
    
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
        
    return result


@router.post("/stats/profiler/{ticker}/filtered", tags=["Stats"])
async def get_profiler_filtered_stats(
    ticker: str,
    payload: dict = Body(...)
):
    """
    Get pre-aggregated profiler stats using filter criteria.
    Payload: {
        "target_session": str,
        "filters": { "Asia": "Short True", ... },
        "broken_filters": { "Asia": "Broken", ... },
        "intra_state": "Any"
    }
    Returns: matched_dates, count, distribution, range_stats
    """
    target_session = payload.get("target_session", "NY1")
    filters = payload.get("filters", {})
    broken_filters = payload.get("broken_filters", {})
    intra_state = payload.get("intra_state", "Any")
    
    result = ProfilerService.get_filtered_stats(
        ticker, target_session, filters, broken_filters, intra_state
    )
    
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result


@router.post("/stats/profiler/{ticker}/price-model", tags=["Stats"])
async def get_profiler_filtered_price_model(
    ticker: str,
    payload: dict = Body(...)
):
    """
    Get Price Model using filter criteria instead of explicit date list.
    Payload: {
        "target_session": str (e.g. "Daily", "NY1"),
        "filters": { "Asia": "Short True", ... },
        "broken_filters": { "Asia": "Broken", ... },
        "intra_state": "Any"
    }
    Returns: average path, extreme path, count
    """
    target_session = payload.get("target_session", "Daily")
    filters = payload.get("filters", {})
    broken_filters = payload.get("broken_filters", {})
    intra_state = payload.get("intra_state", "Any")
    
    result = ProfilerService.get_filtered_price_model(
        ticker, target_session, filters, broken_filters, intra_state
    )
    
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result

@router.get("/stats/hod-lod/{ticker}", tags=["Stats"])
async def get_hod_lod_stats(ticker: str):
    """
    Get pre-computed HOD/LOD time statistics.
    """
    json_path = DATA_DIR / f"{ticker}_hod_lod.json"
    
    if not json_path.exists():
        raise HTTPException(status_code=404, detail=f"HOD/LOD data for {ticker} not found. Run precompute script.")
    
    with open(json_path, 'r') as f:
        data = json.load(f)
    
    return data

@router.get("/stats/range-dist/{ticker}", tags=["Stats"])
async def get_range_distribution(ticker: str):
    """
    Get pre-computed price range distribution (high/low relative to open).
    """
    json_path = DATA_DIR / f"{ticker}_range_dist.json"
    
    if not json_path.exists():
        raise HTTPException(status_code=404, detail=f"Range distribution for {ticker} not found. Run precompute script.")
    
    with open(json_path, 'r') as f:
        data = json.load(f)
    
    return data

@router.post("/stats/clear-cache/{ticker}", tags=["Stats"])
async def clear_profiler_cache(ticker: str = "NQ1"):
    """
    Clear the in-memory cache for profiler data.
    This forces the server to reload from the JSON file on next request.
    """
    from api.services.profiler_service import ProfilerService
    return ProfilerService.clear_cache(ticker)

@router.post("/stats/clear-cache", tags=["Stats"])
async def clear_all_profiler_cache():
    """Clear all in-memory cache."""
    from api.services.profiler_service import ProfilerService
    return ProfilerService.clear_cache()

@router.get("/stats/daily-hod-lod/{ticker}", tags=["Stats"])
async def get_daily_hod_lod(ticker: str):
    """
    Get pre-computed true daily HOD/LOD times (from 1-minute data).
    Returns dict mapping date -> {hod_time, lod_time, hod_price, lod_price, ...}
    """
    json_path = DATA_DIR / f"{ticker}_daily_hod_lod.json"
    
    if not json_path.exists():
        raise HTTPException(status_code=404, detail=f"Daily HOD/LOD data for {ticker} not found. Run precompute_daily_hod_lod.py")
    
    with open(json_path, 'r') as f:
        data = json.load(f)
    
    return data

@router.get("/stats/level-touches/{ticker}", tags=["Stats"])
async def get_level_touches(ticker: str):
    """
    Get pre-computed reference level touch data (PDH/PDL/PDM, P12 H/L/M).
    Returns dict mapping date -> {pdh: {level, touched, touch_time}, ...}
    """
    json_path = DATA_DIR / f"{ticker}_level_touches.json"
    
    if not json_path.exists():
        raise HTTPException(status_code=404, detail=f"Level touch data for {ticker} not found. Run precompute_level_touches.py")
    
    with open(json_path, 'r') as f:
        data = json.load(f)
    
    return data

@router.get("/stats/reference", tags=["Stats"])
async def get_reference_stats():
    """
    Get Reference Data (aggregated stats and medians) from docs folder.
    """
    docs_dir = DATA_DIR.parent / "docs"
    ref_all_path = docs_dir / "ReferenceAll.json"
    ref_med_path = docs_dir / "ReferenceMedian.json"
    
    if not ref_all_path.exists() or not ref_med_path.exists():
        raise HTTPException(status_code=404, detail="Reference data not found in docs/ directory.")
        
    with open(ref_all_path, 'r') as f:
        ref_all = json.load(f)
        
    with open(ref_med_path, 'r') as f:
        ref_med = json.load(f)
        
    return {
        "stats": ref_all,
        "median": ref_med
    }

@router.get("/stats/price-model/{ticker}", tags=["Stats"])
async def get_price_model(
    ticker: str,
    session: str,
    outcome: str,
    days: int = Query(50)
):
    """
    Get Price Model (Composite High/Low) for a specific outcome.
    Returns Average and Extreme models.
    """
    result = ProfilerService.get_price_model_data(ticker, session, outcome, days)
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result

@router.post("/stats/price-model/custom", tags=["Stats"])
async def get_custom_price_model(
    payload: dict = Body(...)
):
    """
    Get Price Model for a specific list of dates (Global Filter Intersection).
    Payload: { "ticker": str, "target_session": str, "dates": List[str] }
    """
    ticker = payload.get("ticker", "NQ1")
    target = payload.get("target_session")
    dates = payload.get("dates", [])
    bucket_minutes = payload.get("bucket_minutes", 1)
    
    result = ProfilerService.get_custom_price_model(ticker, target, dates, bucket_minutes)
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result

# ============================================================================
# NEW: Filter-Based Endpoints (Architecture Refactor)
# ============================================================================

@router.post("/stats/filtered-stats", tags=["Stats"])
async def get_filtered_profiler_stats(
    payload: dict = Body(...)
):
    """
    Get pre-aggregated profiler stats using filter criteria.
    Payload: {
        "ticker": str,
        "target_session": str,
        "filters": { "Asia": "Short True", ... },
        "broken_filters": { "Asia": "Broken", ... },
        "intra_state": "Any"
    }
    Returns: matched_dates, count, distribution, range_stats
    """
    ticker = payload.get("ticker", "NQ1")
    target_session = payload.get("target_session", "NY1")
    filters = payload.get("filters", {})
    broken_filters = payload.get("broken_filters", {})
    intra_state = payload.get("intra_state", "Any")
    
    result = ProfilerService.get_filtered_stats(
        ticker, target_session, filters, broken_filters, intra_state
    )
    
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result


@router.post("/stats/filtered-price-model", tags=["Stats"])
async def get_filtered_price_model(
    payload: dict = Body(...)
):
    """
    Get Price Model using filter criteria instead of explicit date list.
    Payload: {
        "ticker": str,
        "target_session": str (price model session, e.g. "Daily", "NY1"),
        "filters": { "Asia": "Short True", ... },
        "broken_filters": { "Asia": "Broken", ... },
        "intra_state": "Any"
    }
    Returns: average path, extreme path, count
    """
    ticker = payload.get("ticker", "NQ1")
    target_session = payload.get("target_session", "Daily")
    filters = payload.get("filters", {})
    broken_filters = payload.get("broken_filters", {})
    intra_state = payload.get("intra_state", "Any")
    bucket_minutes = payload.get("bucket_minutes", 1)
    
    result = ProfilerService.get_filtered_price_model(
        ticker, target_session, filters, broken_filters, intra_state, bucket_minutes
    )
    
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result

File: web/lib/api/profiler.ts

import { API_BASE_URL } from '../config';

export interface ProfilerSession {
    date: string;
    session: string;
    open: number;
    range_high: number;
    range_low: number;
    mid: number;
    high_time: string | null;
    low_time: string | null;
    high_pct: number;
    low_pct: number;
    close_pct?: number; // Added for Path Distribution
    status: string;
    status_time: string | null;
    broken: boolean;
    broken_time: string | null;
    start_time: string;
    end_time: string;
}

export interface ProfilerResponse {
    sessions: ProfilerSession[];
    metadata: {
        ticker: string;
        days: number;
        count: number;
    };
}



// HOD/LOD Types
export interface TimeStats {
    median: string | null;
    mode: string | null;
    count: number;
    distribution: Record<string, number>;
}

export interface SessionTimeStats {
    high_stats: TimeStats;
    low_stats: TimeStats;
}

export interface HodLodResponse {
    daily: {
        hod_times: string[];  // Raw times for dynamic rebucketing
        lod_times: string[];
        hod_stats: TimeStats;
        lod_stats: TimeStats;
    };
    sessions: Record<string, SessionTimeStats>;
    metadata: {
        ticker: string;
        generated_at: string;
    };
}

export async function fetchHodLodStats(ticker: string): Promise<HodLodResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/hod-lod/${ticker}`);
    if (!res.ok) {
        throw new Error('Failed to fetch HOD/LOD stats');
    }
    return res.json();
}

// Range Distribution Types
export interface RangeDistStats {
    median: number | null;
    mode: number | null;
    count: number;
    distribution: Record<string, number>;
}

export interface RangeDistSession {
    high: RangeDistStats;
    low: RangeDistStats;
    range: RangeDistStats;
}

export interface RangeDistResponse {
    daily: RangeDistSession;
    sessions: Record<string, RangeDistSession>;
    metadata: {
        ticker: string;
        generated_at: string;
    };
}

export async function fetchRangeDist(ticker: string): Promise<RangeDistResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/range-dist/${ticker}`);
    if (!res.ok) {
        throw new Error('Failed to fetch range distribution');
    }
    return res.json();
}

// Daily HOD/LOD times (true daily high/low times from 1-minute data)
export interface DailyHodLodEntry {
    hod_time: string;  // HH:MM format
    lod_time: string;
    hod_price: number;
    lod_price: number;
    daily_high: number;
    daily_low: number;
    daily_open: number;
}

export type DailyHodLodResponse = Record<string, DailyHodLodEntry>;

export async function fetchDailyHodLod(ticker: string): Promise<DailyHodLodResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/daily-hod-lod/${ticker}`);
    if (!res.ok) {
        throw new Error('Failed to fetch daily HOD/LOD data');
    }
    return res.json();
}

// Level touch data (PDH/PDL/PDM, P12 H/L/M touch times)
export interface LevelTouchEntry {
    level: number;
    touched: boolean;
    hits: Record<string, string>; // session name -> first hit time (HH:MM)
    // touch_times: string[]; REMOVED for optimization
}

export interface DayLevelTouches {
    pdh: LevelTouchEntry;
    pdl: LevelTouchEntry;
    pdm: LevelTouchEntry;
    p12h?: LevelTouchEntry;
    p12l?: LevelTouchEntry;
    p12m?: LevelTouchEntry;
    // Time-based opens
    daily_open?: LevelTouchEntry;
    midnight_open?: LevelTouchEntry;
    open_0730?: LevelTouchEntry;
    // Session mids
    asia_mid?: LevelTouchEntry;
    london_mid?: LevelTouchEntry;
    ny1_mid?: LevelTouchEntry;
    ny2_mid?: LevelTouchEntry;
}

export type LevelTouchesResponse = Record<string, DayLevelTouches>;

export async function fetchLevelTouches(ticker: string): Promise<LevelTouchesResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/level-touches/${ticker}?t=${new Date().getTime()}`);
    if (!res.ok) {
        throw new Error('Failed to fetch level touches data');
    }
    return res.json();
}

export interface PriceModelEntry {
    time_idx: number; // Minutes from session start
    time?: string;    // HH:MM time label from server
    high: number;     // % from Open
    low: number;      // % from Open
}

export interface PriceModelResponse {
    median: PriceModelEntry[];
    extreme: PriceModelEntry[];
    count: number;
}



// ============================================================================
// NEW: Filter-Based API Functions (Server-Side Filtering)
// ============================================================================

export interface FilterPayload {
    ticker: string;
    target_session: string;
    filters: Record<string, string>;
    broken_filters: Record<string, string>;
    intra_state: string;
    bucket_minutes?: number; // Optional bucket size (1, 5, 15)
}

export interface FilteredStatsResponse {
    matched_dates: string[];
    count: number;
    distribution: Record<string, number>;
    range_stats: {
        high_pct: Record<string, number>;
        low_pct: Record<string, number>;
    };
    sessions: ProfilerSession[];  // Filtered sessions returned by server
    target_session: string;
    filters_applied: Record<string, string>;
    broken_filters_applied: Record<string, string>;
}

/**
 * Fetch pre-aggregated profiler stats using filter criteria.
 * Filtering is done on the server for better performance.
 */
export async function fetchFilteredStats(payload: FilterPayload): Promise<FilteredStatsResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/filtered-stats`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            ticker: payload.ticker,
            target_session: payload.target_session,
            filters: payload.filters,
            broken_filters: payload.broken_filters,
            intra_state: payload.intra_state
        })
    });
    if (!res.ok) {
        throw new Error('Failed to fetch filtered stats');
    }
    return res.json();
}

/**
 * Fetch Price Model using filter criteria (server-side filtering).
 * Returns average and extreme price paths for matching dates.
 */
export async function fetchFilteredPriceModel(payload: FilterPayload): Promise<PriceModelResponse> {
    const res = await fetch(`${API_BASE_URL}/stats/filtered-price-model`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            ticker: payload.ticker,
            target_session: payload.target_session,
            filters: payload.filters,
            broken_filters: payload.broken_filters,
            intra_state: payload.intra_state,
            bucket_minutes: payload.bucket_minutes
        })
    });
    if (!res.ok) {
        throw new Error('Failed to fetch filtered price model');
    }
    return res.json();
}

// Level Stats (Probability & Timing)
export interface LevelStat {
    rate: number;
    median: number;
    mode: number;
    avg_rel: number; // calculated in precompute
    count: number;
    hits: number;
}

export interface LevelContextData {
    PDH: LevelStat;
    PDL: LevelStat;
    GlobexOpen: LevelStat;
    MidnightOpen: LevelStat;
    AsiaMid: LevelStat;
    LondonMid: LevelStat;
    NY1Mid: LevelStat;
}

export interface AllLevelStats {
    All: LevelContextData;
    Green: LevelContextData;
    Red: LevelContextData;
}

export async function fetchLevelStats(ticker: string): Promise<AllLevelStats> {
    const res = await fetch(`${API_BASE_URL}/stats/profiler/${ticker}/levels`);
    if (!res.ok) {
        throw new Error('Failed to fetch level stats');
    }
    return res.json();
}
